File: springboot-jwt/src/main/java/com/grepp/spring/infra/auth/oauth2/user/OAuth2FailureHandler.java
package com.grepp.spring.infra.auth.oauth2.user;

import com.grepp.spring.app.model.auth.AuthService;
import com.grepp.spring.app.model.auth.token.RefreshTokenService;
import com.grepp.spring.infra.auth.token.JwtProvider;
import com.grepp.spring.infra.auth.token.code.TokenType;
import io.jsonwebtoken.Claims;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.SimpleUrlAuthenticationSuccessHandler;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class OAuth2FailureHandler extends SimpleUrlAuthenticationSuccessHandler {

    private final AuthService authService;
    private final JwtProvider jwtProvider;
    private final RefreshTokenService refreshTokenService;

    @Override
    public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response,
        Authentication authentication) throws IOException, ServletException {

        String requestAccessToken = jwtProvider.resolveToken(request, TokenType.ACCESS_TOKEN);
        String requestRefreshToken = jwtProvider.resolveToken(request, TokenType.REFRESH_TOKEN);

        if (requestAccessToken == null && requestRefreshToken == null) {
            return;
        }

        Claims claims = jwtProvider.parseClaim(requestAccessToken);
        refreshTokenService.deleteByAccessTokenId(claims.getId());
    }
}

File: grepp-gateway/src/main/kotlin/com/grepp/spring/infra/security/oauth2/OAuth2FailureHandler.kt
package com.grepp.spring.infra.security.oauth2

import com.grepp.spring.token.code.TokenType
import com.grepp.spring.token.service.RefreshTokenService
import com.grepp.spring.token.util.JwtProvider
import com.grepp.spring.token.util.TokenResponseExecutor
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler
import org.springframework.stereotype.Component

@Component
class OAuth2FailureHandler(
    private val jwtProvider: JwtProvider,
    private val refreshTokenService: RefreshTokenService
): SimpleUrlAuthenticationFailureHandler() {

    override fun onAuthenticationFailure(
        request: HttpServletRequest,
        response: HttpServletResponse,
        exception: AuthenticationException
    ) {

        val requestAccessToken = jwtProvider.resolveToken(request, TokenType.ACCESS_TOKEN) ?: return
        val claims = jwtProvider.parseClaim(requestAccessToken) ?: return

        refreshTokenService.deleteByAccessTokenId(claims.id)
        TokenResponseExecutor.clear(response)
        redirectStrategy.sendRedirect(request, response, "/")
    }
}